### PR TITLE
OAuth: Allow assigning Server Admin

### LINKED ISSUE #29211: OAuth assign Server Admin Permission
**What would you like to be added**:
Grafana `v7.3` provides option to automatically assign **Grafana Server Admin** permission to specific **LDAP Group**. But this is not possible if you want to use **OAuth** authentication.

**Why is this needed**:
It is only possible to map roles *(e.g. Admin, Viewer)*, but it has less privilege than **Grafana Server Admin**.

So the feature request is about to provide **OAuth** option to automatically assign **Grafana Server Admin** permission like we currently have for roles.

**References**:
Grafana Server Admin - https://grafana.com/docs/grafana/latest/permissions/#grafana-server-admin

LDAP to configure Grafana Server Admin `grafana_admin` option - https://grafana.com/docs/grafana/latest/auth/ldap/

OAuth Role mapping - no Grafana Server Admin option - https://grafana.com/docs/grafana/latest/auth/generic-oauth/#role-mapping

### PR DESCRIPTION
**Pull request by Jguer**
**What this PR does / why we need it**:

- Allow assigning Server Admin (using `GrafanaAdmin` as group/role) 
- `allow_assign_grafana_admin` defaulting to false 
- Remove unused or duplicated attributes
- Unify gitlab/github/okta group extraction logic
- Unified errors

Overview slides: https://docs.google.com/presentation/d/1DqR58DT4faykjaKnkWGsQlXsD-7FDUr4eoVfrWy_Zkw/edit#slide=id.gd42b955d29_0_775

**Which issue(s) this PR fixes**:

Closes https://github.com/grafana/grafana/issues/29211

**Special notes for your reviewer**:

### CODE REVIEW

File: docs/sources/setup-grafana/configure-security/configure-authentication/azuread.md [hunk 1]
```diff
@@ -100,6 +100,30 @@ To enable the Azure AD OAuth2, register your application with Azure AD.
 
 1. Click on **Users and Groups** and add Users/Groups to the Grafana roles by using **Add User**.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/azuread.md:101]: This step confuses me a bit. Is the user clicking **Users and Groups** and then clicking **Add User**? I'm not sure what you mean by 'using Add User'.

File: docs/sources/setup-grafana/configure-security/configure-authentication/azuread.md [hunk 2]
```diff
@@ -100,6 +100,30 @@ To enable the Azure AD OAuth2, register your application with Azure AD.
...
 
+### Assign server administrator privileges
+
+> Available in Grafana v9.2 and later versions.
+
+If the application role received by Grafana is `GrafanaAdmin`, the user will be granted server administrator privileges.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/azuread.md:107]:
```suggestion
If the application role received by Grafana is `GrafanaAdmin`, Grafana grants the user server administrator privileges.  
```

File: docs/sources/setup-grafana/configure-security/configure-authentication/azuread.md [hunk 3]
```diff
@@ -100,6 +100,30 @@ To enable the Azure AD OAuth2, register your application with Azure AD.
 
 1. Click on **Users and Groups** and add Users/Groups to the Grafana roles by using **Add User**.
 
+### Assign server administrator privileges
+
+> Available in Grafana v9.2 and later versions.
+
+If the application role received by Grafana is `GrafanaAdmin`, the user will be granted server administrator privileges.  
+This is useful if you want to grant server administrator privileges to a subset of users.  
+The user will also be assigned the role of `Admin` of the default organization.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/azuread.md:109]: This is passive voice. Who or what is doing the assigning? In my suggestion, I am making an assumption that the 'actor' is Grafana.

Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/azuread.md:109]:
```suggestion
Grafana also assigns the user the `Admin` role of the default organization.
```

File: docs/sources/setup-grafana/configure-security/configure-authentication/azuread.md [hunk 4]
```diff
@@ -100,6 +100,30 @@ To enable the Azure AD OAuth2, register your application with Azure AD.
...
+
+The setting `allow_assign_grafana_admin` under `[auth.azuread]` must be set to `true` for this to work.  
+If the setting is set to `false`, the user will be assigned the role of `Admin` of the default organization, but not server administrator privileges.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/azuread.md:112]:
```suggestion
If the setting is set to `false`, the user is assigned the role of `Admin` of the default organization, but not server administrator privileges.
```

File: docs/sources/setup-grafana/configure-security/configure-authentication/generic-oauth.md [hunk 5]
```diff
@@ -296,6 +296,27 @@ Config:
 role_attribute_path = contains(info.roles[*], 'admin') && 'Admin' || contains(info.roles[*], 'editor') && 'Editor' || 'Viewer'
 ```
 
+#### Map server administrator privileges
+
+> Available in Grafana v9.2 and later versions.
+
+If the role mapping expression evaluates to `GrafanaAdmin`, the user will be granted server administrator privileges.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/generic-oauth.md:303]:
```suggestion
If the role mapping expression evaluates to `GrafanaAdmin`, Grafana grants the user server administrator privileges.  
```

File: docs/sources/setup-grafana/configure-security/configure-authentication/generic-oauth.md [hunk 6]
```diff
@@ -296,6 +296,27 @@ Config:
 role_attribute_path = contains(info.roles[*], 'admin') && 'Admin' || contains(info.roles[*], 'editor') && 'Editor' || 'Viewer'
 ```
 
+#### Map server administrator privileges
+
+> Available in Grafana v9.2 and later versions.
+
+If the role mapping expression evaluates to `GrafanaAdmin`, the user will be granted server administrator privileges.  
+This is useful if you want to grant server administrator privileges to a subset of users.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/generic-oauth.md:304]: See above suggestion

File: docs/sources/setup-grafana/configure-security/configure-authentication/generic-oauth.md [hunk 7]
```diff
@@ -296,6 +296,27 @@ Config:
 role_attribute_path = contains(info.roles[*], 'admin') && 'Admin' || contains(info.roles[*], 'editor') && 'Editor' || 'Viewer'
 ```
 
+#### Map server administrator privileges
+
+> Available in Grafana v9.2 and later versions.
+
+If the role mapping expression evaluates to `GrafanaAdmin`, the user will be granted server administrator privileges.  
+This is useful if you want to grant server administrator privileges to a subset of users.  
+The user will also be assigned the role of `Admin` of the default organization.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/generic-oauth.md:305]: See suggestion above.

File: docs/sources/setup-grafana/configure-security/configure-authentication/generic-oauth.md [hunk 8]
```diff
@@ -296,6 +296,27 @@ Config:
...
+
+The setting `allow_assign_grafana_admin` under `[auth.generic_oauth]` must be set to `true` for this to work.  
+If the setting is set to `false`, the user will be assigned the role of `Admin` of the default organization, but not server administrator privileges.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/generic-oauth.md:308]:
```suggestion
If the setting is set to `false`, the user is assigned the role of `Admin` of the default organization, but not server administrator privileges.
```

File: docs/sources/setup-grafana/configure-security/configure-authentication/github.md [hunk 9]
```diff
@@ -130,6 +130,27 @@ role_attribute_path = contains(groups[*], '@github/example-group') && 'Editor' |
 
 Note: If a match is found in other fields, teams will be ignored.
 
+#### Map server administrator privileges
+
+> Available in Grafana v9.2 and later versions.
+
+If the role mapping expression evaluates to `GrafanaAdmin`, the user will be granted server administrator privileges.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/github.md:137]: see suggestion above.

File: docs/sources/setup-grafana/configure-security/configure-authentication/github.md [hunk 10]
```diff
@@ -130,6 +130,27 @@ role_attribute_path = contains(groups[*], '@github/example-group') && 'Editor' |
 
 Note: If a match is found in other fields, teams will be ignored.
 
+#### Map server administrator privileges
+
+> Available in Grafana v9.2 and later versions.
+
+If the role mapping expression evaluates to `GrafanaAdmin`, the user will be granted server administrator privileges.  
+This is useful if you want to grant server administrator privileges to a subset of users.  
+The user will also be assigned the role of `Admin` of the default organization.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/github.md:139]: see suggestion above.

File: docs/sources/setup-grafana/configure-security/configure-authentication/github.md [hunk 11]
```diff
@@ -130,6 +130,27 @@ role_attribute_path = contains(groups[*], '@github/example-group') && 'Editor' |
...
+
+The setting `allow_assign_grafana_admin` under `[auth.github]` must be set to `true` for this to work.  
+If the setting is set to `false`, the user will be assigned the role of `Admin` of the default organization, but not server administrator privileges.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/github.md:142]:
```suggestion
If the setting is set to `false`, the user is assigned the role of `Admin` of the default organization, but not server administrator privileges.
```

File: docs/sources/setup-grafana/configure-security/configure-authentication/gitlab.md [hunk 12]
```diff
@@ -139,12 +144,33 @@ Groups can also be used to map roles. Group name (lowercased and unique) is used
...
 role_attribute_path = contains(groups[*], 'example-group') && 'Editor' || 'Viewer'
 ```
 
 Note: If a match is found in other fields, groups will be ignored.
 
+#### Map server administrator privileges
+
+> Available in Grafana v9.2 and later versions.
+
+If the role mapping expression evaluates to `GrafanaAdmin`, the user will be granted server administrator privileges.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/gitlab.md:157]: same suggestion as above.

File: docs/sources/setup-grafana/configure-security/configure-authentication/gitlab.md [hunk 13]
```diff
@@ -139,12 +144,33 @@ Groups can also be used to map roles. Group name (lowercased and unique) is used
...
 
 Note: If a match is found in other fields, groups will be ignored.
 
+#### Map server administrator privileges
+
+> Available in Grafana v9.2 and later versions.
+
+If the role mapping expression evaluates to `GrafanaAdmin`, the user will be granted server administrator privileges.  
+This is useful if you want to grant server administrator privileges to a subset of users.  
+The user will also be assigned the role of `Admin` of the default organization.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/gitlab.md:159]: see suggestion above.

File: docs/sources/setup-grafana/configure-security/configure-authentication/gitlab.md [hunk 14]
```diff
@@ -139,12 +144,33 @@ Groups can also be used to map roles. Group name (lowercased and unique) is used
...
+
+The setting `allow_assign_grafana_admin` under `[auth.gitlab]` must be set to `true` for this to work.  
+If the setting is set to `false`, the user will be assigned the role of `Admin` of the default organization, but not server administrator privileges.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/gitlab.md:162]:
```suggestion
If the setting is set to `false`, the user is assigned the role of `Admin` of the default organization, but not server administrator privileges.
```

File: docs/sources/setup-grafana/configure-security/configure-authentication/okta.md [hunk 15]
```diff
@@ -77,6 +77,27 @@ Grafana uses JSON obtained from querying the `/userinfo` endpoint for the path l
...
+#### Map server administrator privileges
+
+> Available in Grafana v9.2 and later versions.
+
+If the role mapping expression evaluates to `GrafanaAdmin`, the user will be granted server administrator privileges.  
+This is useful if you want to grant server administrator privileges to a subset of users.  
+The user will also be assigned the role of `Admin` of the default organization.
+
+The setting `allow_assign_grafana_admin` under `[auth.okta]` must be set to `true` for this to work.  
+If the setting is set to `false`, the user will be assigned the role of `Admin` of the default organization, but not server administrator privileges.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/okta.md:89]:
```suggestion
If the setting is set to `false`, the user is assigned the role of `Admin` of the default organization, but not server administrator privileges.
```

File: docs/sources/setup-grafana/configure-security/configure-authentication/okta.md [hunk 16]
```diff
@@ -77,6 +77,27 @@ Grafana uses JSON obtained from querying the `/userinfo` endpoint for the path l
 
 Read about how to [add custom claims](https://developer.okta.com/docs/guides/customize-tokens-returned-from-okta/add-custom-claim/) to the user info in Okta. Also, check Generic OAuth page for [JMESPath examples]({{< relref "generic-oauth/#jmespath-examples" >}}).
 
+#### Map server administrator privileges
+
+> Available in Grafana v9.2 and later versions.
+
+If the role mapping expression evaluates to `GrafanaAdmin`, the user will be granted server administrator privileges.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-security/configure-authentication/okta.md:84]: see suggestion above.

**Review by chri2547 — COMMENTED** [review]
Added suggestions and one question.

**Review by gamab — APPROVED** [review]
I like that you have added uniformity amongst oauth providers, for instance `okta_oauth` can now leverage groups to assign roles just like the others :+1:
I'm wondering if we might get complains about people wanting a "Grafana Admin Viewer" for instance.
Only edge case I thought of, but believe it doesn't make much sense, is: `allow_assign_grafana_admin = false` and `role_attribute_path = is_admin && 'GrafanaAdmin' || 'Viewer'`, the user will be an Admin in such a case. But that's a configuration error that should not happen IMO.

File: pkg/login/social/generic_oauth.go [hunk 17]
```diff
@@ -147,12 +144,18 @@ func (s *SocialGenericOAuth) UserInfo(client *http.Client, token *oauth2.Token)
 		}
 
 		if userInfo.Role == "" {
-			role, err := s.extractRole(data)
-			if err != nil {
-				s.log.Warn("Failed to extract role", "error", err)
-			} else if role != "" {
+			role, grafanaAdmin := s.extractRoleAndAdmin(data.rawJSON, []string{})
+			if role != "" {
+				if s.roleAttributeStrict && !role.IsValid() {
```
[the remark this answers is not in the document]
Comment (Jguer) [reply pkg/login/social/generic_oauth.go:149]: this keeps legacy behavior intact, we could change it in another PR :)

File: docs/sources/setup-grafana/configure-security/configure-authentication/azuread.md (hunk 1 above)
Comment (Jguer) [reply docs/sources/setup-grafana/configure-security/configure-authentication/azuread.md:101]: this is not related to this PR and no change was applied here
.
I've checked Azure's UI and this is indeed what the author of the line meant: The user clicks Users and Groups and then clicks Add User once they arrive on the new menu

**Pull request merged**